content/docs/user-guide/project-structure/pipelines-files.md

```diff
@@ -0,0 +1,416 @@
+> incompatible with `dvc run`.
+
+Let's say we have `params.yaml` (default params file) with the following
+contents:
+
+```yaml
+models:
+  us:
+    threshold: 10
+    filename: 'model-us.hdf5'
+```
+
+Those values can be used anywhere in `dvc.yaml` with the `${}` _substitution
+expression_:
+
+<!-- prettier-ignore-start -->
+```yaml
+stages:
+  build-us:
+    cmd: >-
+      python train.py
+      --thresh ${models.us.threshold}
+      --out ${models.us.filename}
+    outs:
+      - ${models.us.filename}:
+        cache: true
+```
+<!-- prettier-ignore-end -->
+
+DVC will track simple param values (numbers, strings, etc.) used in `${}` (they
+will be listed by `dvc params diff`).
+
+Alternatively, values for substitution can be listed as top-level `vars` like
+this:
+
+```yaml
+vars:
+  - models:
+      us:
+        threshold: 10
+  - desc: 'Reusable description'
+
+stages:
+  build-us:
+    desc: ${desc}
+    cmd: python train.py --thresh ${models.us.threshold}
+```
+
+> DVC merges values from params files and `vars` when possible. For example,
+> `{"grp": {"a": 1}}` merges with `{"grp": {"b": 2}}`, but not with
+> `{"grp": {"a": 7}}`.
+
+> Note that values from `vars` are not tracked like parameters.
+
+To load additional params files, list them in the top `vars`, in the desired
+order, e.g.:
+
+> Their paths will be evaluated based on
+> [`wdir`](/doc/user-guide/dvc-files/dvc.yaml#accepted-fields), if one given.
+
+```yaml
+vars:
+  - params.json
+  - myvar: 'value'
+  - config/myapp.yaml
+```
+
+(ℹ️) Note that the default `params.yaml` file is always loaded first, if
+present.
+
+It's also possible to specify what to include from additional params files, with
+a `:` colon:
+
+```yaml
+vars:
+  - params.json:clean,feats
+
+stages:
+  featurize:
+    cmd: ${feats.exec}
+    deps:
+      - ${clean.filename}
+    outs:
+      - ${feats.dirname}
+```
+
+Stage-specific values are also supported, with inner `vars`. You may also load
+additional params files locally. For example:
+
+```yaml
+stages:
+  build-us:
+    vars:
+      - params.json:build
+      - model:
+          filename: 'model-us.hdf5'
+    cmd: python train.py ${build.epochs} --out ${model.filename}
+    outs:
+      - ${model.filename}
+```
+
+⚠️ Known limitations of local `vars`:
+
+- [`wdir`](/doc/user-guide/dvc-files/dvc.yaml#accepted-fields) cannot use values
+  from local `vars`, as DVC uses the working directory first (to load any values
+  from params files listed in `vars`).
+- `foreach` is also incompatible with local `vars` at the moment.
+
+The substitution expression supports these forms:
+
+```yaml
+${param} # Simple
+${param.key} # Nested values through . (period)
+${param.list[0]} # List elements via index in [] (square brackets)
+```
+
+> To use the expression in `dvc.yaml`, for example for the shell to interpret a
+> command substitution, escape it with a backslash, e.g. `\${...`.
+
+## Generating multiple stages
```

Comment:
still don't like the title ... let's brainstorm a bit ...

Reply:
+1. Throwing out other title ideas:

* Looping over stages
* `foreach` stages

Reply:
I've been opposed to the "loop" analogy from the beginning here haha. But it has been discussed (see iterative/dvc/issues/5312) so at this point if that's the consensus, I have no strong opinion against. "Foreach-stages" or maybe "Stage loops" do sound user-friendly.